### PR TITLE
Add feature for access control

### PR DESCRIPTION
**Pull request by panpan0721**
There are two sample features of access control.

The one feature provides in access_control_api are three apis which
reporting access control target, checking the access rights and checking
nonce as the proxy of access control server abilities.

The proxy has two env configs:
EHSM_CONFIG_DEFENSE_ENABLE //The total key for access control feature.
Default config value is empty or set it to "false", no access control
taking effect and checking nonce will using inline code.

EHSM_CONFIG_DEFENSE_SERVER_URL //This is the server address for restful
APIs and the key for local or remote access control feature.
This config needs EHSM_CONFIG_DEFENSE_ENABLE setting "true" first.
Default config value is empty and using local access control feature by
integrating access control functions in ehsm kms server.
Setting remote URL will enable remote access control feature, and all
accessing functions will be redirected to remote access control server.
The sample of config:
//No feature:

Config nothing or
export EHSM_CONFIG_DEFENSE_ENABLE="false"

//Local Mode:
export EHSM_CONFIG_DEFENSE_ENABLE="true"

//Remote Mode:
export EHSM_CONFIG_DEFENSE_ENABLE="true"
export EHSM_CONFIG_DEFENSE_SERVER_URL="http://<access_srv_address:access_srv_port>"

The other one is a sample of access control server.
All access control functions have local mode and remote mode.
In local mode, the ehsm kms server will integrating access control
functions as inline code.
In remote mode, there is a standalone ehsm access server providing access
control feature by exposing three restful API as:
POST <access_srv_address>/access?Action=CheckAccess
POST <access_srv_address>/access?Action=ReportTarget
POST <access_srv_address>/access?Action=CheckNonce

This feature has four env configs as
  EHSM_CONFIG_DENY_SIZE_LIMIT //The max deny Ip list size, default value is 3000

  EHSM_CONFIG_DENY_TIMEOUT //The max deny time for one Ip.
  The unit is MS and default value is 1800000 MS(30 min)

  EHSM_CONFIG_DENY_COUNT_LIMIT //The max deny count for one Ip.
  If one Ip has been reported max count within the configured timeout.
  The default value is 5

  EHSM_CONFIG_REPORTERS // A List for reporters.
  This server will allow the reporters in config reporting deny Ips.
  The empty config means allowing anyone to report deny Ip.
  The default value is empty.

The configs also have local mode and remote mode.
In local mode, the config EHSM_CONFIG_REPORTERS will not take effect and
all configs need config into ehsm kms server env.

In remote mode, all configs taking effect and need config into ehsm access
server env.

The sample of configs:
//Local Mode (EHSM KMS Server Env)
EHSM_CONFIG_DENY_SIZE_LIMIT=1000
EHSM_CONFIG_DENY_TIMEOUT=10000
EHSM_CONFIG_DENY_COUNT_LIMIT=3

//Remote Mode (EHSM Access Server Env)
EHSM_CONFIG_DENY_SIZE_LIMIT=1000
EHSM_CONFIG_DENY_TIMEOUT=10000
EHSM_CONFIG_DENY_COUNT_LIMIT=3
EHSM_CONFIG_REPORTERS="<reporter Ip 1>|<reporter Ip 2>"

Signed-off-by: panpan0721 <2271409777@qq.com>

### CODE REVIEW

File: ehsm_kms_service/access_control_server.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    result: {
+      ...data,
+    },
+  }
+}
+
+function DenyListClearUp(timestamp) {
+  let _size_limit = parseInt(EHSM_CONFIG_DENY_SIZE_LIMIT)
+  if (isNaN(_size_limit)) {
+    _size_limit = 3000
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:35]: define default macros

File: ehsm_kms_service/function.js [hunk 2]
```diff
@@ -500,6 +504,7 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
           return
```
Comment (yang8621) [inline ehsm_kms_service/function.js:504]: ReportControlTarget(req.ip, new Date().getTime()) at above line 499/503.

Comment (yang8621) [reply ehsm_kms_service/function.js:504]: ReportControlTarget(req.ip, new Date().getTime()) at above line 439/442.

File: ehsm_kms_service/access_control_server.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  }
+  let _time_out = parseInt(EHSM_CONFIG_DENY_TIMEOUT)
+  if (isNaN(_time_out)) {
+    _time_out = 1800000 //ms defult 30 min
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:39]: define default macros

File: ehsm_kms_service/access_control_server.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    ip_from = ip_from.substr(7)
+  }
+  var current = new Date().getTime()
+  DenyListClearUp(current)
+  switch (action) {
+    case defense_apis.CheckAccess:
+      try {
+        let count_limit = parseInt(EHSM_CONFIG_DENY_COUNT_LIMIT)
+        if (isNaN(count_limit)) {
+          count_limit = 5
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:83]: define default macros

File: ehsm_kms_service/access_control_server.js (hunk 1 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_server.js]: It will be fixed

File: ehsm_kms_service/access_control_server.js (hunk 3 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_server.js]: It will be fixed

File: ehsm_kms_service/access_control_server.js (hunk 4 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_server.js]: It will be fixed

File: ehsm_kms_service/access_control_server.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  var current = new Date().getTime()
+  DenyListClearUp(current)
+  switch (action) {
+    case defense_apis.CheckAccess:
+      try {
+        let count_limit = parseInt(EHSM_CONFIG_DENY_COUNT_LIMIT)
+        if (isNaN(count_limit)) {
+          count_limit = default_config_value.count_limit
+        }
+        let accessDeny = (!_DenyList.has(ip) || (_DenyList.get(ip).count < count_limit))
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:91]: The name accessDeny is confusing.  Actually, it should be accessible, right?
rename to _access.

File: ehsm_kms_service/function.js (hunk 2 above)
Comment (panpan0721) [reply ehsm_kms_service/function.js:509]: It will be fixed

**Comment by yang8621** [comment]
fix the repos in commit-msg: witch

File: ehsm_kms_service/access_control_server.js (hunk 5 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_server.js:91]: It will be fixed

File: ehsm_kms_service/k8s/ehsm-kms.yaml [hunk 6]
```diff
@@ -178,6 +187,131 @@ spec:
...
+      protocol: TCP
+      # This port is [ehsm access control service] access port,you can change it to what you want.
+      port: 9010
+      targetPort: 9010
+      nodePort: 30010
+  sessionAffinity: ClientIP
+  externalIPs:
+  # This IP is [ehsm access control service] access IP, You need to adjust 5.6.7.8 to the service IP you want to expose.
+  - 5.6.7.8
+
```
Comment (syan10) [inline ehsm_kms_service/k8s/ehsm-kms.yaml:314]: there's no difference with the ehsm-kms service? does it means the external user could also access it?

Comment (wanghouqi) [reply ehsm_kms_service/k8s/ehsm-kms.yaml:314]: It can be accessed in the LAN

File: ehsm_kms_service/function.js [hunk 7]
```diff
@@ -444,71 +450,91 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
...
+      const nonce_data = { nonce, nonce_timestamp: new Date().getTime() }
+      if (!nonce_database[appid]) {
+        nonce_database[appid] = [nonce_data]
+      } else if (
+        !!nonce_database[appid] &&
+        nonce_database[appid].findIndex(
+          (nonce_data) => nonce_data.nonce == nonce
+        ) > -1
+      ) {
+        res.send(_result(400, "Timestamp can't be repeated in 20 minutes"))
```
Comment (yang8621) [inline ehsm_kms_service/function.js:523]: ReportControlTarget(req.ip, new Date().getTime())

File: ehsm_kms_service/function.js [hunk 8]
```diff
@@ -444,71 +450,91 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
...
+        nonce_database[appid].unshift(nonce_data)
+      }
+      _lastSteps()
+    } else {
+      _checkNonce.then((checkNonceRes) => {
+        if (checkNonceRes) {
+          _lastSteps()
         } else {
-          next()
+          res.send(_result(400, "Timestamp can't be repeated in 20 minutes"))
```
Comment (yang8621) [inline ehsm_kms_service/function.js:534]: ReportControlTarget(req.ip, new Date().getTime())

File: ehsm_kms_service/access_control_server.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+  if (!nonce_database[appid]) {
+    nonce_database[appid] = [nonce_data]
+    return true;
+  }
+
+  if (nonce_database[appid].findIndex(
+      (nonce_data) => nonce_data.nonce == nonce
+    ) > -1
+  ) {
+    return false
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:123]: DefenseReportTarget

File: ehsm_kms_service/access_control_server.js [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+      })
+  }
+}
+const DefenseCheckNonce = (nonce_database, appid, nonce, nonce_timestamp) => {
+  const nonce_data = {
+    nonce: nonce,
+    nonce_timestamp: nonce_timestamp
+  }
+
+  if (!nonce_database[appid]) {
```
Comment (yang8621) [inline ehsm_kms_service/access_control_server.js:114]: not found the logic to delete the nonce after it's expired, e.g.
https://github.com/intel/ehsm/blob/main/ehsm_kms_service/function.js#L54

File: ehsm_kms_service/function.js (hunk 7 above)
Comment (panpan0721) [reply ehsm_kms_service/function.js:523]: This part is no access control feature and should not report any. 
Do you mean adding nonce repeated failed into access control behavior？

File: ehsm_kms_service/function.js (hunk 8 above)
Comment (panpan0721) [reply ehsm_kms_service/function.js:534]: This is same as before.

File: ehsm_kms_service/access_control_server.js (hunk 10 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_server.js:114]: Yes. We used _nonce_cache_timer() in function.js.
You can see it at ehsm_kms_server.js line 32 and ehsm_access_server.js line 16

File: ehsm_kms_service/common_function.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
+const logger = require('./logger')
```
Comment (yang8621) [inline ehsm_kms_service/common_function.js:1]: Nice

File: ehsm_kms_service/ehsm_kms_server.env [hunk 12]
```diff
@@ -0,0 +1,14 @@
+#ehsm_kms_server
+export EHSM_CONFIG_COUCHDB_USERNAME=admin
+export EHSM_CONFIG_COUCHDB_PASSWORD=password
+export EHSM_CONFIG_COUCHDB_SERVER=10.112.240.122
+export EHSM_CONFIG_COUCHDB_PORT=5984
+export EHSM_CONFIG_COUCHDB_DB=ehsm_kms_db2
```
Comment (yang8621) [inline ehsm_kms_service/ehsm_kms_server.env:6]: what is this DB used for?  
Both nonce and deny_list in the control server are based-on memory.

File: ehsm_kms_service/access_control_api.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+const {
+  EHSM_CONFIG_DEFENSE_ENABLE,
+  EHSM_CONFIG_DEFENSE_SERVER_URL,
+} = process.env
+
+const _DefenseType = {
+  NONE: 1,
+  LOCAL: 2,
+  REMOTE: 3
+}
```
Comment (syan10) [inline ehsm_kms_service/access_control_api.js:20]: does it need to distinguish the remote and lcoal?
not sure if the below works:
Disabled -- not set the URL
Enabled -- set a valid URL (a ocal host IP or remote IP)

so, the EHSM_CONFIG_DEFENSE_ENABLE should be deleted.

File: ehsm_kms_service/ehsm_access_server.env [hunk 14]
```diff
@@ -0,0 +1,5 @@
+#ehsm_access_server
+export EHSM_CONFIG_REPORTERS=""
+export EHSM_CONFIG_DENY_SIZE_LIMIT=2
```
Comment (yang8621) [inline ehsm_kms_service/ehsm_access_server.env:3]: does 2 make sense?
add some #comment to each env. e.g.
EHSM_CONFIG_REPORTERS:  The whitelisted IP who can report violations to access the server.

File: ehsm_kms_service/function.js (hunk 7 above)
Comment (yang8621) [reply ehsm_kms_service/function.js:454]: OK.  I know.   This part is the branch that EHSM_CONFIG_DEFENSE_ENABLE is false or defenseType is None, right?

File: ehsm_kms_service/function.js [hunk 15]
```diff
@@ -444,71 +381,92 @@ const _checkParams = function (req, res, next, nonce_database, DB) {
...
-    } else {
-      nonce_database[appid].unshift(nonce_data)
-    }
 
-    if (ACTION !== key_management_apis[ACTION]) {
-      // check payload
-      const _checkPayload_res = _checkPayload(req, res, next)
-      if (!_checkPayload_res) {
-        return
+    var _lastSteps = () => {
```
Comment (yang8621) [inline ehsm_kms_service/function.js:399]: Not understand

File: ehsm_kms_service/access_control_api.js (hunk 13 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_api.js:20]: image

File: ehsm_kms_service/function.js (hunk 15 above)
Comment (panpan0721) [reply ehsm_kms_service/function.js:399]: _checkParamsis a async function, but this function _checkParams is not a async function.
There should wait the _checkParams result but could not use "await".
So I make the last steps of checking params as a function for async call.

File: ehsm_kms_service/ehsm_access_server.env (hunk 14 above)
Comment (panpan0721) [reply ehsm_kms_service/ehsm_access_server.env:3]: This file is just for local test run. I will delete it.

File: ehsm_kms_service/ehsm_kms_server.env (hunk 12 above)
Comment (panpan0721) [reply ehsm_kms_service/ehsm_kms_server.env:6]: This file is just for local test run. I will delete it.

File: ehsm_kms_service/access_control_api.js (hunk 13 above)
Comment (panpan0721) [reply ehsm_kms_service/access_control_api.js:20]: And also local mode is realized by calling function, a local host IP is invalid in docker

**Comment by syan10** [comment]
close it due to currently can't get the right IP address for the target machine.

**Pull request closed without merging**